### PR TITLE
Lesson03 mailroom.py updated with bubblesort and unnecessary return r…

### PR DESCRIPTION
**Pull request by wpucilow**
…emoved

Updated version of mailroom.py
- removed unnecessary "return"
- removed unused main_menu_1()
- added bubblesort
- report sorted by Total donation (descending order)

### CODE REVIEW

File: students/Wieslaw_Pucilowski/lesson03/mailroom.py [hunk 1]
```diff
@@ -85,18 +72,29 @@ def send_thank_you():
...
+            madeSwap = False
+            for j in range(0,len(ListOfLists)-1):
+                if sum(ListOfLists[j][1:]) > sum(ListOfLists[j+1][1:]):
+                    temp = ListOfLists[j+1]
+                    ListOfLists[j+1] = ListOfLists[j]
+                    ListOfLists[j] = temp
+                    madeSwap = True
+            if not madeSwap:
+                return ListOfLists
+        return ListOfLists
```
Comment (natasha-aleksandrova) [inline students/Wieslaw_Pucilowski/lesson03/mailroom.py]: @wpucilow couple comments 🙂 
* you have mixed spacing here.. make sure to _only_ use 4 space indent
* per pep8 style guidelines, var names should follow snake_case convention
* you should use `sorted` method to sort your donors, no need to implement your own sorting logic; please check out this doc (sorted + itemgetter combo): https://wiki.python.org/moin/HowTo/Sorting

**Comment by wpucilow** [comment]
Hi Natasha,
My initial solution was
def sort_sub_list(sub_li, rev=False):
    rev=rev
    sub_li.sort(key = lambda x: sum(x[1:]), reverse=rev)
    return sub_li
    
def create_report():
    print("Creating a report...")
    print( "{:<30}| {:<18}| {:<8}| {:<18}".format('Donor Name','Total Given','Num Gifts','Average Gift'))
    print("{:-<80}".format(''))
    for i in sort_sub_list(donors_hist, True):
        print('{:<30}{}{:>18.2f}{:>11}{}{:>17.2f}'.format(i[:1][0], ' $', sum(i[1:]), len(i[1:]), ' $', sum(i[1:])/len(i[1:])))
    return
but I have changed it to bubblesort because I'm not sure about lambda. I know it works and I have used it in the past put it was not mention during the course. In there any explanation about the lambda?
Wieslaw

**Comment by natasha-aleksandrova** [comment]
you don't need to know lambdas for this :)

in the link I provided there is an example on how to use sorted + itemgetter:

`sorted(donors, key=itemgetter(0))`

**Comment by wpucilow** [comment]
Yes, but can't do sorted(donors, key=itemgetter(sum(donors[1:])))
so I though about the following, creating index=sum(donors[1:]) in the copy of the donorsand then sorting by this index
def create_report_sort():
    print("Creating a report...")
    report=donors_hist[:]
    for x in report:
        x.insert(0,sum(x[1:]))
    print( "{:<30}| {:<18}| {:<8}| {:<18}".format('Donor Name','Total Given','Num Gifts','Average Gift'))
    print("{:-<80}".format(''))
    for i in sorted(report, key=itemgetter(0), reverse=True):
        print('{:<30}{}{:>18.2f}{:>11}{}{:>17.2f}'.format(i[:1][0], ' $', i[0], len(i[2:]), ' $', i[0]/len(i[2:])))
    return
Wieslaw

**Comment by natasha-aleksandrova** [comment]
@wpucilow note that you can comment on github directly and also format your code so that it is easier to read.

you have two options:
* create a list with summary data (that includes your avg, total etc) and sort that
* or use custom sort key like below example
```
>>> def custom_key(item):
...     return sum(item[1])
...
>>> sorted(donors, key=custom_key)
[['Jack', [4]], ['Bob', [100, 200]]]
>>> sorted(donors, key=custom_key, reverse=True)
[['Bob', [100, 200]], ['Jack', [4]]]```

**Comment by wpucilow** [comment]
Yes, it seems to be the best solution.

def custom_key(a):
    return sum(a[1:])

def create_report():
    print("Creating a report...\n")
    print( "{:<30}| {:<18}| {:<8}| {:<18}".format('Donor Name','Total Given','Num Gifts','Average Gift'))
    print("{:-<80}".format(''))
    for x in sorted(donors_hist, key=custom_key,reverse=True):
        print('{:<30}{}{:>18.2f}{:>11}{}{:>17.2f}'.format(x[:1][0], ' $', sum(x[1:]), len(x[1:]), ' $', sum(x[1:])/len(x[1:])))

**Comment by wpucilow** [comment]
Hi Natasha,
mailroom.py has been updated and new pull request
Wieslaw

**Pull request merged**